File: main2.py
import numpy as np
import matplotlib.pyplot as plt
from PIL import Image

# Load and preprocess the image
def load_image(image_path):
    image = Image.open(image_path).convert('L')  # Convert to grayscale
    image = image.resize((256, 256))  # Resize to 256x256
    image_array = np.array(image, dtype=float)
    return image_array

# Jacobi method for solving the Laplace equation
def jacobi_method(image, max_iter=100, tol=1e-4):
    u = image.copy()
    u_new = np.zeros_like(u)
    error = np.inf
    iteration = 0

    while error > tol and iteration < max_iter:
        u_new[1:-1, 1:-1] = 0.25 * (u[1:-1, :-2] + u[1:-1, 2:] + u[:-2, 1:-1] + u[2:, 1:-1])
        error = np.linalg.norm(u_new - u) / np.linalg.norm(u)
        u = u_new.copy()
        iteration += 1

    print(f"Jacobi converged in {iteration} iterations with error {error}")
    return u

# Gauss-Seidel method for solving the Laplace equation
def gauss_seidel_method(image, max_iter=100, tol=1e-4):
    u = image.copy()
    error = np.inf
    iteration = 0

    while error > tol and iteration < max_iter:
        u_old = u.copy()
        for i in range(1, u.shape[0]-1):
            for j in range(1, u.shape[1]-1):
                u[i, j] = 0.25 * (u[i-1, j] + u[i+1, j] + u[i, j-1] + u[i, j+1])
        error = np.linalg.norm(u - u_old) / np.linalg.norm(u_old)
        iteration += 1

    print(f"Gauss-Seidel converged in {iteration} iterations with error {error}")
    return u

# Visualize results
def visualize_results(original, jacobi_result, gauss_seidel_result):
    plt.figure(figsize=(15, 5))
    
    plt.subplot(1, 3, 1)
    plt.imshow(original, cmap='gray')
    plt.title('Original Image')
    
    plt.subplot(1, 3, 2)
    plt.imshow(jacobi_result, cmap='gray')
    plt.title('Jacobi Method')
    
    plt.subplot(1, 3, 3)
    plt.imshow(gauss_seidel_result, cmap='gray')
    plt.title('Gauss-Seidel Method')
    
    plt.show()

# Main function
def main():
    # Load the image
    image_path = 'lena.png'  # Replace with your image path
    original_image = load_image(image_path)
    
    # Apply Jacobi method
    jacobi_result = jacobi_method(original_image)
    
    # Apply Gauss-Seidel method
    gauss_seidel_result = gauss_seidel_method(original_image)
    
    # Visualize the results
    visualize_results(original_image, jacobi_result, gauss_seidel_result)

if __name__ == "__main__":
    main()

File: main2.m
% MATLAB Implementation for Edge Detection using Laplace Operator

% Step 1: Load and preprocess the image
image_path = 'lena.png'; % Replace with your image path
original_image = imread(image_path);
original_image = rgb2gray(original_image); % Convert to grayscale if needed
original_image = imresize(original_image, [256, 256]); % Resize to 256x256
original_image = double(original_image); % Convert to double for numerical computations

% Step 2: Apply the Laplace operator for edge detection
laplace_filter = [0 1 0; 1 -4 1; 0 1 0]; % Laplace kernel
edges = imfilter(original_image, laplace_filter, 'same'); % Apply Laplace filter

% Step 3: Normalize and threshold the edges
edges = abs(edges); % Take absolute value to highlight edges
edges = edges / max(edges(:)); % Normalize to [0, 1]
edges = edges > 0.1; % Apply a threshold to binarize the edges

% Step 4: Visualize the results
figure;
subplot(1, 2, 1);
imshow(uint8(original_image)); % Display original image
title('Original Image');

subplot(1, 2, 2);
imshow(edges); % Display edge-detected image
title('Edge Detection (Laplace Operator)');